File: memory-domain/src/main/java/com/memory/domain/file/File.java
package com.memory.domain.file;

import com.memory.domain.BaseTimeEntity;
import com.memory.domain.member.Member;
import com.memory.domain.memory.Memory;
import jakarta.persistence.*;
import lombok.*;

@ToString
@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class File extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String originalFileName;

    private String fileName;

    private String fileUrl;

    @Enumerated(EnumType.STRING)
    private FileType fileType;

    private Long fileSize;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "memory_id")
    private Memory memory;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    private Member member;

    @Builder
    public File(String originalFileName, String fileName, String fileUrl, FileType fileType, Long fileSize) {
        this.originalFileName = originalFileName;
        this.fileName = fileName;
        this.fileUrl = fileUrl;
        this.fileType = fileType;
        this.fileSize = fileSize;
    }

    public void updateMemory(Memory memory) {
        this.memory = memory;
    }

    public void updateMember(Member member) {
        this.member = member;
    }

    public boolean validateMember(Long memberId) {
        return this.getMember() != null && !this.getMember().getId().equals(memberId);
    }
}

File: memory-domain/src/main/kotlin/com/memory/domain/file/File.kt
package com.memory.domain.file

import com.memory.domain.BaseTimeEntity
import com.memory.domain.member.Member
import com.memory.domain.memory.Memory
import jakarta.persistence.*

@Entity
@Table(name = "file")
class File(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    var originalFileName: String,

    var fileName: String,

    var fileUrl: String,

    @Enumerated(EnumType.STRING)
    var fileType: FileType,

    var fileSize: Long,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "memory_id")
    var memory: Memory? = null,

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    var member: Member? = null
) : BaseTimeEntity() {

    constructor(originalFileName: String, fileName: String, fileUrl: String, fileType: FileType, fileSize: Long) : this(
        id = null,
        originalFileName = originalFileName,
        fileName = fileName,
        fileUrl = fileUrl,
        fileType = fileType,
        fileSize = fileSize
    )

    fun updateMemory(memory: Memory?) {
        this.memory = memory
    }

    fun updateMember(member: Member?) {
        this.member = member
    }

    fun validateMember(memberId: Long): Boolean {
        return this.member != null && this.member!!.id != memberId
    }
}
